File: example/java/jimmer-sql-graphql/src/main/java/org/babyfish/jimmer/sql/example/graphql/cfg/CommonEntityDraftInterceptor.java
package org.babyfish.jimmer.sql.example.graphql.cfg;

import org.babyfish.jimmer.ImmutableObjects;
import org.babyfish.jimmer.sql.DraftInterceptor;
import org.babyfish.jimmer.sql.example.graphql.entities.CommonEntityDraft;
import org.babyfish.jimmer.sql.example.graphql.entities.CommonEntityProps;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;

/*
 * see JSqlClient.Builder.addDraftInterceptors
 */
@Component
public class CommonEntityDraftInterceptor implements DraftInterceptor<CommonEntityDraft> {

    @Override
    public void beforeSave(CommonEntityDraft draft, boolean isNew) {
        if (!ImmutableObjects.isLoaded(draft, CommonEntityProps.MODIFIED_TIME)) {
            draft.setModifiedTime(LocalDateTime.now());
        }
        if (isNew && !ImmutableObjects.isLoaded(draft, CommonEntityProps.CREATED_TIME)) {
            draft.setCreatedTime(LocalDateTime.now());
        }
    }
}

File: example/kotlin/jimmer-sql-graphql-kt/src/main/kotlin/org/babyfish/jimmer/example/kt/graphql/cfg/CommonEntityDraftInterceptor.kt
package org.babyfish.jimmer.example.kt.graphql.cfg

import org.babyfish.jimmer.example.kt.graphql.entities.CommonEntity
import org.babyfish.jimmer.example.kt.graphql.entities.CommonEntityDraft
import org.babyfish.jimmer.kt.isLoaded
import org.babyfish.jimmer.sql.DraftInterceptor
import org.springframework.stereotype.Component
import java.time.LocalDateTime

/*
 * see KSqlClientDsl.addDraftInterceptors
 */
@Component
class CommonEntityDraftInterceptor : DraftInterceptor<CommonEntityDraft> {

    override fun beforeSave(draft: CommonEntityDraft, isNew: Boolean) {
        if (!isLoaded(draft, CommonEntity::modifiedTime)) {
            draft.modifiedTime = LocalDateTime.now()
        }
        if (isNew && !isLoaded(draft, CommonEntity::createdTime)) {
            draft.createdTime = LocalDateTime.now()
        }
    }
}